sssd_test_framework/topology_controllers.py

```diff
@@ -79,10 +202,42 @@ class IPATrustADTopologyController(BackupTopologyController):
     IPA trust AD Topology Controller.
     """
 
-    pass
+    @restore_vanilla_on_error
+    def topology_setup(self, client: ClientHost, ipa: IPAHost, trusted: ADHost | SambaHost) -> None:
+        if self.provisioned:
+            self.logger.info(f"Topology '{self.name}' is already provisioned")
+            return
+
+        # Create trust
+        self.logger.info(f"Establishing trust between {ipa.domain} and {trusted.domain}")
+        ipa.kinit()
+        ipa.ssh.exec(
+            ["ipa", "trust-add", trusted.domain, "--admin", "Administrator", "--password"],
+            input=trusted.adminpw,
+        )
+
+        # Do not enroll client into IPA domain if it is already joined
```

Comment:
Why is this here and not for AD join?

Reply:
Can you clarify this question?

Reply:
In `IPATrustADTopologyController` topology_setup method you have:

```
        # Do not enroll client into IPA domain if it is already joined
        if "ipa" not in self.multihost.provisioned_topologies:

```

But this same check does not exist in `ADTopologyController`, where both execute 'realm join'

Reply:
This is different. This handles the case where the client is already joined into an IPA domain (from ansible), but trust between IPA and AD is not yet created. Which is the default setup on containers. So in this case, you only want to create the trust between IPA and AD but client is already enrolled into IPA domain.

In ADTopologyController, you just want to joint the client into AD.

Reply:
I understand now, thanks for clarification.